sound/soc/sof/intel/hda-loader.c

```diff
@@ -238,28 +238,29 @@ static int cl_cleanup(struct snd_sof_dev *sdev, struct snd_dma_buffer *dmab,
 static int cl_copy_fw(struct snd_sof_dev *sdev, struct hdac_ext_stream *stream)
 {
 	unsigned int reg;
-	int ret, status;
+	int ret;
 
 	ret = cl_trigger(sdev, stream, SNDRV_PCM_TRIGGER_START);
 	if (ret < 0) {
 		dev_err(sdev->dev, "error: DMA trigger start failed\n");
 		return ret;
 	}
 
-	status = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR,
+	ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_BAR,
 					HDA_DSP_SRAM_REG_ROM_STATUS, reg,
 					((reg & HDA_DSP_ROM_STS_MASK)
 						== HDA_DSP_ROM_FW_ENTERED),
 					HDA_DSP_REG_POLL_INTERVAL_US,
 					HDA_DSP_BASEFW_TIMEOUT_US);
 
+	if (ret < 0)
+		return ret;
+
```

Comment:
@plbossart I think the original flow was intentional. Irrespective of whether status is successful or not, we need to stop the DMA

Reply:
It's unclear if there is actually a design intent, there are two levels of error substitution with the following code in the  caller

````c
	ret = cl_copy_fw(sdev, stream);
	if (!ret)
		dev_dbg(sdev->dev, "Firmware download successful, booting...\n");
	else
		dev_err(sdev->dev, "error: load fw failed ret: %d\n", ret);

cleanup:
	/*
	 * Perform codeloader stream cleanup.
	 * This should be done even if firmware loading fails.
	 */
	ret1 = cl_cleanup(sdev, &sdev->dmab, stream);
	if (ret1 < 0) {
		dev_err(sdev->dev, "error: Code loader DSP cleanup failed\n");

		/* set return value to indicate cleanup failure */
		ret = ret1;
	}
````

so status returned by the poll_timeout maybe be replaced by the cl_trigger_stop return value or the cleanup value. Yay.

Reply:
@plbossart yes I agree that this is messy. But I think the intention is that the dma be stopped and codeloader stream cleaned up irrespective of whether copy was successful or not. Eventually, I think we should return the first error value from all of three steps.

Reply:
ok, fixed now